src/main/scala/ai/diffy/proxy/TimedMulticastService.scala

```diff
@@ -2,13 +2,37 @@ package ai.diffy.proxy
 
 import com.twitter.finagle.Service
 import com.twitter.util.{Future, Try}
+import com.twitter.finagle.http.Request
 
-class TimedMulticastService[-A, +B](services: Seq[Service[A, B]])
+case class Server(classifier: Int)
+
+class TimedMulticastService[-A, +B, C](services: Seq[Service[A, B]], apiRoots : Seq[C])
   extends Service[A, Seq[(Try[B], Long, Long)]]
 {
+  var requestCount = 0
+  var dest = ""
+
+  def addApiRoot(server: Server, request: Request): Unit = {
+    val apiRoot = server match {
+      case Server(0) => apiRoots(0).toString
+      case Server(1) => apiRoots(1).toString
+      case Server(2) => apiRoots(2).toString
+    }
+    request.uri(dest + apiRoot)
+  }
+
   def apply(request: A): Future[Seq[(Try[B], Long, Long)]] =
     services.foldLeft[Future[Seq[(Try[B], Long, Long)]]](Future.Nil){ case (acc, service) =>
       acc flatMap { responseTriesWithTiming =>
+        if (request.isInstanceOf[Request]) {
+          if (dest.equals(""))
+            dest = request.asInstanceOf[Request].uri
+          addApiRoot(Server(requestCount), request.asInstanceOf[Request])
+          requestCount += 1
+          if (requestCount == 3)
+            requestCount = 0
```

Comment:
This doesn't work when you have multiple (100s, 100s or even 10s of) requests being concurrently proxied. What you really need here is to zip the services with apiRoots and only do that mapping once at setup time rather than repeat the work on each request. Doing so also eliminates unnecessary concept clutter.